test/python/circuit/test_controlled_gate.py

```diff
@@ -0,0 +1,402 @@
+from qiskit.quantum_info.operators.predicates import matrix_equal, is_unitary_matrix
+import qiskit.circuit.add_control as ac
+from qiskit.transpiler.passes import Unroller
+from qiskit.converters.circuit_to_dag import circuit_to_dag
+from qiskit.converters.dag_to_circuit import dag_to_circuit
+from qiskit.quantum_info import Operator
+from qiskit.extensions.standard import (CnotGate, XGate, YGate, ZGate, U1Gate,
+                                        CyGate, CzGate, Cu1Gate, SwapGate)
+from qiskit.extensions.unitary import UnitaryGate
+
+
+@ddt
+class TestControlledGate(QiskitTestCase):
+    """ControlledGate tests."""
+
+    def test_controlled_x(self):
+        """Test creation of controlled x gate"""
+        self.assertEqual(XGate().q_if(), CnotGate())
+
+    def test_controlled_y(self):
+        """Test creation of controlled x gate"""
+        self.assertEqual(YGate().q_if(), CyGate())
+
+    def test_controlled_z(self):
+        """Test creation of controlled x gate"""
+        self.assertEqual(ZGate().q_if(), CzGate())
+
+    def test_controlled_u1(self):
+        """Test creation of controlled x gate"""
+        theta = 0.5
+        self.assertEqual(U1Gate(theta).q_if(), Cu1Gate(theta))
+
+    def test_circuit_append(self):
+        """Test appending controlled gate to quantum circuit"""
+        circ = QuantumCircuit(5)
+        inst = CnotGate()
+        circ.append(inst.q_if(), qargs=[0, 2, 1])
+        circ.append(inst.q_if(2), qargs=[0, 3, 1, 2])
+        circ.append(inst.q_if().q_if(), qargs=[0, 3, 1, 2])  # should be same as above
+        self.assertEqual(circ.depth(), 3)
+        self.assertEqual(circ[0][0].num_ctrl_qubits, 2)
+        self.assertEqual(circ[1][0].num_ctrl_qubits, 2)
+        self.assertEqual(circ[2][0].num_ctrl_qubits, 1)
+        self.assertEqual(circ[0][0].num_qubits, 3)
+        self.assertEqual(circ[1][0].num_qubits, 4)
+        self.assertEqual(circ[2][0].num_qubits, 4)
+        for instr in circ:
+            gate = instr[0]
+            self.assertTrue(isinstance(gate, ControlledGate))
+
+    def test_definition_specification(self):
+        """Test instantiation with explicit definition"""
+        swap = SwapGate()
+        cswap = ControlledGate('cswap', 3, [], num_ctrl_qubits=1,
+                               definition=swap.definition)
+        self.assertEqual(swap.definition, cswap.definition)
+
+    def test_multi_controlled_composite_gate(self):
+        """Test multi controlled composite gate"""
+        num_ctrl = 3
+        ctrl_dim = 2**num_ctrl
+        # create composite gate
+        sub_q = QuantumRegister(2)
+        cgate = QuantumCircuit(sub_q, name='cgate')
+        cgate.h(sub_q[0])
+        cgate.crz(np.pi/2, sub_q[0], sub_q[1])
+        cgate.swap(sub_q[0], sub_q[1])
+        cgate.u3(0.1, 0.2, 0.3, sub_q[1])
+        cgate.t(sub_q[0])
+        num_target = cgate.width()
+        gate = instruction_to_gate(cgate.to_instruction())
```

Comment:
maybe add a `inst.to_gate()` shorthand, similar to the `circ.to_instruction()` shorthand?

Reply:
Perhaps this can be handled in a follow up PR.